aten/src/ATen/native/Blas.cpp

```diff
@@ -137,6 +140,110 @@ Tensor dot(const Tensor &self, const Tensor &other){
   });
 }
 
+// this is to port LDA_COND from the TH codebase
+// #define LDA_COND(M, N, LDA) ((N) == 1 || (LDA) >= THMax(1, (M)))
+// lda_cond function in this file is not used
+// because it requires `lda > std::max<int64_t>(1L, m)`
+// instead of `>=` as defined in the TH implementation
+constexpr inline bool lda_cond_addr(int64_t m, int64_t n, int64_t lda) {
+  return n == 1 || lda >= std::max<int64_t>(1L, m);
+}
+
+template<typename scalar_t>
+Tensor &addr_impl_cpu(Tensor &result, const Tensor &self,
+                    const Tensor& vec1, const Tensor& vec2,
+                    scalar_t beta, scalar_t alpha) {
+  if (&result != &self) {
+    at::native::resize_as_(result, self);
```

Comment:
Part of the reason I'd like to consider writing this op as a composite of other ops is to avoid dealing with logic like this or maintaining it as it changes. Here, for example, result is not the same size as self (maybe that's handled correctly later, I'm still reading through). Also, if we want addr to participate in type promotion we'll have to implement it for addr, but if we implement addr in terms of its component ops then we should "inherit" their type promotion behavior.

Reply:
Thanks for pointing this out, implementing `addr` as a composite op makes the code much more maintainable.